ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -0,0 +1,110 @@
+import com.github.ambry.quota.QuotaUtils;
+import com.github.ambry.rest.RestRequest;
+import com.mysql.cj.PreparedQuery;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link QuotaEnforcer} for Capacity Units of Ambry resource.
+ */
+public class AmbryCUQuotaEnforcer implements QuotaEnforcer {
+  private static final Logger logger = LoggerFactory.getLogger(AmbryCUQuotaEnforcer.class);
+
+  private final QuotaSource quotaSource;
+  private final float maxFrontendCuUsageToAllowExceed;
+
+  /**
+   * Constructor for {@link AmbryCUQuotaEnforcer}.
+   * @param quotaSource {@link QuotaSource} where the quota limit and usage will be saved and retrieved from.
+   */
+  public AmbryCUQuotaEnforcer(QuotaSource quotaSource, float maxFrontendCuUsageToAllowExceed) {
+    this.quotaSource = quotaSource;
+    this.maxFrontendCuUsageToAllowExceed = maxFrontendCuUsageToAllowExceed;
+  }
+
+  @Override
+  public void init() {
+  }
+
+  @Override
+  public QuotaRecommendation chargeAndRecommend(RestRequest restRequest, BlobInfo blobInfo,
+      Map<QuotaName, Double> requestCostMap) {
+    return runWithNoException(() -> {
+      if (!quotaSource.isReady()) {
+        return null;
+      }
+      if (requestCostMap.size() == 0) {
+        logger.warn("Empty cost map provided for {} request for blob {}. Nothing to charge",
+            (restRequest == null ? "null" : restRequest.getRestMethod().name()),
+            (blobInfo == null ? "null" : blobInfo.getBlobProperties()));
+      }
+      quotaSource.charge(restRequest, blobInfo, requestCostMap);
+      return quotaSource.checkResourceUsage(restRequest);
+    }, restRequest, "chargeAndRecommend", createServeRecommendation(restRequest));
+  }
+
+  @Override
+  public QuotaRecommendation getResourceRecommendation(RestRequest restRequest) {
+    return runWithNoException(() -> {
+      if (!quotaSource.isReady()) {
+        return null;
+      }
+      return quotaSource.checkResourceUsage(restRequest);
+    }, restRequest, "getResourceRecommendation", createServeRecommendation(restRequest));
+  }
+
+  @Override
+  public boolean isQuotaExceedAllowed(RestRequest restRequest) {
+    return runWithNoException(() -> {
+      QuotaRecommendation feQuotaRecommendation = quotaSource.checkFrontendUsage(restRequest);
+      return feQuotaRecommendation.getQuotaUsagePercentage() < maxFrontendCuUsageToAllowExceed;
+    }, restRequest, "isQuotaExceedAllowed", true);
+  }
+
+  @Override
+  public QuotaSource getQuotaSource() {
+    return quotaSource;
+  }
+
+  @Override
+  public void shutdown() {
+
+  }
+
+  private QuotaRecommendation createServeRecommendation(RestRequest restRequest) {
+    return new QuotaRecommendation(false, -1, QuotaUtils.getQuotaName(restRequest), 200, -1);
+  }
+
+  private <T> T runWithNoException(Callable<T> callable, RestRequest restRequest, String methodName,
```

Comment:
You might want to add the request request in the warning message? Or remove the restRequest if it's not used.